Earthfile

```diff
@@ -39,241 +59,326 @@ setup:
+###############################################################################
+prepare-image:
+  ARG --required VCS_REF
+
+  FROM +pre-installation
 
-  # Specify the docker image metadata
   LABEL org.label-schema.schema-version="1.0"
   LABEL org.label-schema.name="Space ROS"
   LABEL org.label-schema.description="Preview version of the Space ROS platform"
   LABEL org.label-schema.vendor="Open Robotics"
   LABEL org.label-schema.url="https://github.com/space-ros"
-  LABEL org.label-schema.vcs-url="https://github.com/space-ros/docker-images"
+  LABEL org.label-schema.vcs-url="https://github.com/space-ros/space-ros"
   LABEL org.label-schema.vcs-ref=${VCS_REF}
 
-  COPY +build/install ${SPACEROS_DIR}/install
-  COPY +spaceros-artifacts/exact.repos ${SPACEROS_DIR}/exact.repos
-  RUN rm -r src
+  # Add missing dependencies
+  RUN apt install -y libspdlog-dev \
+      python3-numpy \
+      python3-rosdistro \
+      tzdata \
+      sudo \
+      python3-colcon-common-extensions \
+      python3-rosdep \
+    && pip3 install pyyaml \
+      lark \
+      packaging \
+      netifaces \
+      catkin_pkg \
+      psutil
+
+  # Prepare the image
+  RUN mkdir -p ${SPACEROS_DIR}
+  COPY +rosdep/rosdeps.sh ${SPACEROS_DIR}/rosdeps.sh
+  RUN bash ${SPACEROS_DIR}/rosdeps.sh
+
+###############################################################################
+### Prepare Image Stage
+# This stage is responsible for preparing the image with the ROS 2 workspace.
+# The image is saved only if the build-test stage is successful.
+###############################################################################
+image:
+  ARG --required IMAGE_VARIANT
+  FROM +prepare-image
+
+  COPY +build/spaceros ${SPACEROS_DIR}
+  COPY +sources/exact.repos ${SPACEROS_DIR}/scripts/spaceros.repos
+  COPY scripts/generate-repos.sh scripts/merge-repos.py ${SPACEROS_DIR}/scripts/
+  RUN chmod +x ${SPACEROS_DIR}/scripts/generate-repos.sh ${SPACEROS_DIR}/scripts/merge-repos.py \
+    && mv ${SPACEROS_DIR}/rosdeps.sh ${SPACEROS_DIR}/scripts/rosdeps.sh
+
+  # Post Installation
+  DO +POST_INSTALLATION --IMAGE_VARIANT=${IMAGE_VARIANT}
 
-  WORKDIR ${HOME_DIR}
-  COPY docker/entrypoint.sh /ros_entrypoint.sh
-  ENTRYPOINT ["/ros_entrypoint.sh"]
-  CMD ["bash"]
-  SAVE IMAGE --push osrf/space-ros:latest osrf/space-ros:$tag
+  # Clear Apt and Pip cache
+  RUN rm -rf /var/lib/apt/lists/* /tmp/pip-reqs /var/cache/apt/archives \
+    && apt-get clean \
+    && pip cache purge
+
+  # Add user and group
+  RUN useradd -m -s /bin/bash ${USERNAME} \
+    && chown -R ${USERNAME}:${USERNAME} ${SPACEROS_DIR}/scripts \
+    && echo "${USERNAME} ALL=(ALL) NOPASSWD:ALL" > /etc/sudoers.d/${USERNAME}
+  RUN echo "source ${SPACEROS_DIR}/setup.bash" >> ${HOME}/.bashrc
+
+  RUN chown -R ${USERNAME}:${USERNAME} ${HOME}
+
+  USER ${USERNAME}
+  WORKDIR ${HOME}
+
+  # Only save image if build-test is successful
+  IF [ "${SKIP_BUILD_TEST}" = "true" ]
+    SAVE IMAGE --push ${IMAGE_NAME}:${IMAGE_VARIANT}
```

Comment:
I'll be really curious to know your thoughts. On `main` I separated the `build` and `push` stages to parameterize the registry that we push to (GHCR for `main` and Dockerhub for tags). See [here](https://github.com/space-ros/space-ros/blob/main/Earthfile#L281). 

Should we do something similar here? Adding a parameterized stage and/or function is much easier with the way you have things setup, and could be very nice for pushing incremental stages for future builds to cache off of. Though for the sake of getting this over the hump, I think keeping parity and pushing the core tagged and main images are sufficient. What do you think?

Reply:
Agreed. Pushing tags to the Dockerhub and the regular updates to GHCR would make it easier to keep track of things.

There are a few extensions to cover in this PR. 
 - Pushing core images to GHCR as main.
 - Pushing core and dev images to Dockerhub on tag releases.
 - This PR should allow parallel CI jobs now which could be addressed as a different issue.

This should help sort out the requirements and developments for the downstream repositories in a better way.